proof/refine/ARM/Finalise_R.thy

```diff
@@ -2723,6 +2779,86 @@ crunch invs[wp]: prepareThreadDelete "invs'"
+  apply (clarsimp simp: ko_wp_at'_def obj_at'_def projectKOs get_refs_def2)
+  done
+
+lemma schedContextUnbindNtfn_obj_at'_ntfnSc:
+  "\<lbrace>valid_objs' and obj_at' (\<lambda>ntfn. ntfnSc ntfn = Some scPtr) ntfnPtr\<rbrace>
+   schedContextUnbindNtfn scPtr
+   \<lbrace>\<lambda>_ s. obj_at' (\<lambda>ntfn. ntfnSc ntfn = None) ntfnPtr s\<rbrace>"
+  apply (simp add: schedContextUnbindNtfn_def sym_refs_asrt_def)
+  apply (wpsimp wp: stateAssert_wp set_ntfn'.obj_at'_strongest getNotification_wp
+                    hoare_vcg_all_lift hoare_vcg_imp_lift')
+  apply (drule ntfnSc_sym_refsD; assumption?)
+  apply (clarsimp simp: obj_at'_def projectKOs)
+  done
+
+lemma schedContextMaybeUnbindNtfn_obj_at'_ntfnSc:
+  "\<lbrace>valid_objs'\<rbrace>
+   schedContextMaybeUnbindNtfn ntfnPtr
+   \<lbrace>\<lambda>_ s. obj_at' (\<lambda>ntfn. ntfnSc ntfn = None) ntfnPtr s\<rbrace>"
+  apply (simp add: schedContextMaybeUnbindNtfn_def)
+  apply (wpsimp wp: schedContextUnbindNtfn_obj_at'_ntfnSc getNotification_wp)
+  apply (clarsimp simp: obj_at'_def)
+  done
+
+lemma replyClear_removeable':
+  "\<lbrace>K (\<exists>can_grant. cap = ReplyCap rptr can_grant)\<rbrace>
+   replyClear rptr tptr
+   \<lbrace>\<lambda>rv s. removeable' slot s cap\<rbrace>"
+  apply (rule hoare_gen_asm_single)
+  apply (simp add: replyClear_def removeable'_def)
+  apply wpsimp
+  apply (clarsimp simp: pred_tcb_at'_def obj_at'_def)
+  done
+
+crunches schedContextUnbindTCB, schedContextCompleteYieldTo
+  for bound_tcb_at'[wp]: "bound_tcb_at' P p"
+  (ignore: threadSet simp: crunch_simps wp: crunch_wps)
+
+lemma schedContextUnbindTCB_valid_queues[wp]:
+  "\<lbrace>valid_queues and valid_objs' and sch_act_simple\<rbrace>
+   schedContextUnbindTCB scPtr
+   \<lbrace>\<lambda>_. valid_queues\<rbrace>"
+  unfolding schedContextUnbindTCB_def
+  apply (wpsimp wp: threadSet_valid_queues tcbReleaseRemove_valid_queues
+                    hoare_vcg_all_lift tcbSchedDequeue_valid_queues
+                    rescheduleRequired_oa_queued haskell_assert_wp
+         | wp (once) hoare_drop_imps)+
+  apply (auto simp: valid_obj'_def valid_sched_context'_def
+              elim: valid_objs'_maxDomain valid_objs'_maxPriority
+             dest!: ko_at_valid_objs'_pre)
+  done
+
+crunches setConsumed
+  for valid_queues[wp]: valid_queues
+  and ksQ[wp]: "\<lambda>s. P (ksReadyQueues s p)"
+  (simp: crunch_simps wp: crunch_wps)
+
+lemma schedContextCompleteYieldTo_valid_queues[wp]:
+  "schedContextCompleteYieldTo tptr \<lbrace>valid_queues\<rbrace>"
+  unfolding schedContextCompleteYieldTo_def
+  apply (wpsimp wp: hoare_vcg_all_lift threadGet_wp simp: inQ_def cong: conj_cong)
+  apply (clarsimp simp: obj_at'_def)
+  done
+
+lemma schedContextUnbindTCB_invs':
+  "\<lbrace>invs' and sch_act_simple and K (scPtr \<noteq> idle_sc_ptr)\<rbrace>
+   schedContextUnbindTCB scPtr
+   \<lbrace>\<lambda>_. invs'\<rbrace>"
+  unfolding schedContextUnbindTCB_def
+  sorry
```

Comment:
Can you please either sort this out before merging or label it with FIXME RT + some comments?